stan/math/prim/prob/wiener5_lpdf.hpp

```diff
@@ -0,0 +1,428 @@
+#ifndef STAN_MATH_PRIM_PROB_WIENER5_LPDF_HPP
+#define STAN_MATH_PRIM_PROB_WIENER5_LPDF_HPP
+
+#include <stan/math/prim/fun.hpp>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+enum class FunType { Density, GradT, GradA, GradV, GradW, GradSV, GradSW };
+
+/**
+ * Implementation function for calculating the density or gradient w.r.t. a
+ * specified parameter of the 5-parameter Wiener distribution
+ *
+ * @tparam FunTypeEnum An enum value indicating the quantity required
+ * @param y A scalar variable; the reaction time in seconds
+ * @param a The boundary separation
+ * @param vn The drift rate
+ * @param wn The relative starting point
+ * @param sv The inter-trial variability of the drift rate
+ * @param err The error tolerance for estimation
+ * @param use_log Whether to return calculation w.r.t. the log value
+ *
+ */
+template <FunType FunTypeEnum>
+inline double wiener5_helper(const double& y, const double& a, const double& vn,
+                             const double& wn, const double& sv,
+                             const double& err = log(1e-12),
+                             bool use_log = false) {
+  double w = 1.0 - wn;
+  double v = -vn;
+  double sv_sqr = square(sv);
+  double one_plus_svsqr_y = 1 + sv_sqr * y;
+  double two_avw = 2 * a * v * w;
+  double two_log_a = 2 * log(a);
+  double ans;
+  double ld = 0;
+
+  if (FunTypeEnum == FunType::GradSV || FunTypeEnum == FunType::GradV) {
+    ld = wiener5_helper<FunType::Density>(y, a, vn, wn, sv, err);
+    if (FunTypeEnum == FunType::GradV) {
+      ans = (a * (1 - wn) - vn * y);
+      if (sv != 0) {
+        ans /= 1 + sv_sqr * y;
+      }
+    } else {
+      double t1 = -y / one_plus_svsqr_y;
+      double t2 = (square(a * w) + two_avw * y + square(v * y))
+                  / square(one_plus_svsqr_y);
+      ans = sv * (t1 + t2);
+    }
+  } else {
+    double lg1;
+    if (sv != 0) {
+      lg1 = (sv_sqr * square(a * w) - two_avw - square(v) * y) / 2.0
+                / one_plus_svsqr_y
+            - two_log_a - 0.5 * log(one_plus_svsqr_y);
+    } else {
+      lg1 = (-two_avw - square(v) * y) / 2.0 - two_log_a;
+    }
+    double ans0 = 0;
+    if (FunTypeEnum != FunType::Density) {
```

Comment:
Break this up into multiple functions. This if statement style is very confusing